File: examples/python/05_collisions.py
# Collisions Demo
# This example connects to an RB-Y1 robot, powers on the specified devices,
# monitors collision distance during motion, and sends a stop command when the robot gets too close. See --help for arguments.
#
# Usage example:
#     python 05_collisions.py --address 192.168.30.1:50051 --model a --power ".*"
#
# Copyright (c) 2025 Rainbow Robotics. All rights reserved.
#
# DISCLAIMER:
# This is a sample code provided for educational and reference purposes only.
# Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
# the use or misuse of this demo code. Please use with caution and at your own discretion.

import argparse
import logging
import threading
import time

import numpy as np
import rby1_sdk as rby

STOP_DISTANCE = 0.03
RIGHT_ARM_TARGET = np.array(
    [0.8, -0.8, 0.7, -2.0, 0, 0.0, 0]
    
)
stop_requested = False



def callback(robot_state):
    global stop_requested
    if not robot_state.collisions:
        return

    nearest = min(robot_state.collisions, key=lambda collision: collision.distance)
    print(
        f"nearest collision distance: {nearest.distance:.4f} | "
        f"{nearest.link1} <-> {nearest.link2}"
    )

    if nearest.distance < STOP_DISTANCE:
        print(f"Collision Detected")
        stop_requested = True

def move_to_pre_control_pose(robot):
    """Move to the pre-control pose before starting the motion."""
    torso = np.array([0.0, 0.1, -0.2, 0.1, 0.0, 0.0])
    right_arm = np.array([0.2, -0.2, 0.0, -1.0, 0, 0.7, 0.0])
    left_arm = np.array([0.2, 0.2, 0.0, -1.0, 0, 0.7, 0.0])
    rv = robot.send_command(
        rby.RobotCommandBuilder().set_command(
            rby.ComponentBasedCommandBuilder().set_body_command(
                rby.BodyComponentBasedCommandBuilder()
                .set_torso_command(
                    rby.JointPositionCommandBuilder()
                    .set_minimum_time(5.0)
                    .set_position(torso)
                )
                .set_right_arm_command(
                    rby.JointPositionCommandBuilder()
                    .set_minimum_time(5.0)
                    .set_position(right_arm)
                )
                .set_left_arm_command(
                    rby.JointPositionCommandBuilder()
                    .set_minimum_time(5.0)
                    .set_position(left_arm)
                )
            )
        ),
        90,
    ).get()
    print(f"pre control pose finish_code: {rv.finish_code}")
    if rv.finish_code != rby.RobotCommandFeedback.FinishCode.Ok:
        exit(1)

def main(address, model, power, servo):
    global stop_requested
    robot = rby.create_robot(address, model)
    if not robot.connect():
        logging.error(f"Failed to connect robot {address}")
        exit(1)
    if not robot.is_power_on(power):
        if not robot.power_on(power):
            logging.error(f"Failed to turn power ({power}) on")
            exit(1)
    if not robot.is_servo_on(servo):
        if not robot.servo_on(servo):
            logging.error(f"Failed to servo ({servo}) on")
            exit(1)
    if robot.get_control_manager_state().state in [
        rby.ControlManagerState.State.MajorFault,
        rby.ControlManagerState.State.MinorFault,
    ]:
        if not robot.reset_fault_control_manager():
            logging.error("Failed to reset control manager")
            exit(1)
    if not robot.enable_control_manager():
        logging.error("Failed to enable control manager")
        exit(1)

    move_to_pre_control_pose(robot)

    stop_requested = False
    robot.start_state_update(callback, rate=50)

    rv = robot.send_command(
        rby.RobotCommandBuilder().set_command(
            rby.ComponentBasedCommandBuilder().set_body_command(
                rby.BodyComponentBasedCommandBuilder()
                .set_right_arm_command(
                    rby.JointPositionCommandBuilder()
                    .set_minimum_time(5.0)
                    .set_position(RIGHT_ARM_TARGET)
                )
            )
        ),
        10,
    )

    while not stop_requested:
        time.sleep(0.01)

    
    robot.cancel_control()
    robot.stop_state_update()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="05_collisions")
    parser.add_argument("--address", type=str, required=True, help="Robot address")
    parser.add_argument(
        "--model", type=str, default="a", help="Robot Model Name (default: 'a')"
    )
    parser.add_argument(
        "--power",
        type=str,
        default=".*",
        help="Power device name regex pattern (default: '.*')",
    )
    parser.add_argument(
        "--servo",
        type=str,
        default=".*",
        help="Servo name regex pattern (default: '.*')",
    )
    args = parser.parse_args()

    main(address=args.address, model=args.model, power=args.power, servo=args.servo)

File: examples/cpp/05_collisions.cpp
// Collisions Demo
// This example is part of the RB-Y1 SDK examples. See --help for arguments.
//
// Usage example:
//   ./example_05_collisions --address 192.168.30.1:50051 --model a --power ".*"
//
// Copyright (c) 2025 Rainbow Robotics. All rights reserved.
//
// DISCLAIMER:
// This is a sample code provided for educational and reference purposes only.
// Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
// the use or misuse of this demo code. Please use with caution and at your own discretion.

#include <algorithm>
#include <atomic>
#include <cctype>
#include <chrono>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <thread>
#include <vector>
#include <Eigen/Dense>
#include "rby1-sdk/model.h"
#include "rby1-sdk/robot.h"

using namespace rb;
using namespace std::chrono_literals;

namespace {

std::atomic<bool> g_stop{false};

void SignalHandler(int) { g_stop = true; }

std::string ToLower(std::string v) {
  std::transform(v.begin(), v.end(), v.begin(), [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return v;
}

std::string FormatEigenVec3(const Eigen::Vector3d& v) {
  Eigen::IOFormat fmt(3, 0, ", ", ", ", "[", "]", "[", "]");
  std::ostringstream ss;
  ss.setf(std::ios::fixed);
  ss << std::setprecision(3) << v.format(fmt);
  return ss.str();
}

std::string CollisionStr(const dyn::CollisionResult& cr) {
  std::ostringstream ss;
  ss << "CollisionResult(link1=" << cr.link1 << ", link2=" << cr.link2
     << ", position1=" << FormatEigenVec3(cr.position1) << ", position2=" << FormatEigenVec3(cr.position2)
     << ", distance=" << std::fixed << std::setprecision(6) << cr.distance << ")";
  return ss.str();
}

void PrintUsage(const char* prog) {
  std::cerr << "Usage: " << prog << " --address <server address> [--model a|m|ub] [--power <regex>]" << std::endl;
  std::cerr << "   or: " << prog << " <server address> [model] [power_regex]" << std::endl;
}

template <typename ModelT>
int RunRobotCollisions(const std::string& address, const std::string& power_regex) {
  auto robot = Robot<ModelT>::Create(address);

  if (!robot->Connect()) {
    std::cerr << "Connection failed" << std::endl;
    return 1;
  }
  if (!robot->IsConnected()) {
    std::cerr << "Robot is not connected" << std::endl;
    return 1;
  }
  if (!robot->IsPowerOn(power_regex)) {
    if (!robot->PowerOn(power_regex)) {
      std::cerr << "Failed to power on" << std::endl;
      return 1;
    }
  }

  robot->StartStateUpdate(
      [](const auto& state) {
        if (!state.collisions.empty()) {
          const auto& collision = state.collisions.front();
          if (collision.distance < 0) {
            std::cout << ">>>>> Collision detected!" << std::endl;
          }
          std::cout << CollisionStr(collision) << std::endl;
        }
      },
      10.0);

  auto deadline = std::chrono::steady_clock::now() + 100s;
  while (!g_stop && std::chrono::steady_clock::now() < deadline) {
    std::this_thread::sleep_for(100ms);
  }

  if (g_stop) {
    std::cout << "Stopping state update..." << std::endl;
  }
  robot->StopStateUpdate();

  return 0;
}

}  // namespace

int main(int argc, char** argv) {
  std::signal(SIGINT, SignalHandler);

  std::string address;
  std::string model = "a";
  std::string power = ".*";

  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg == "--address" && i + 1 < argc) {
      address = argv[++i];
    } else if (arg == "--model" && i + 1 < argc) {
      model = argv[++i];
    } else if (arg == "--power" && i + 1 < argc) {
      power = argv[++i];
    } else if (arg.rfind("--", 0) == 0) {
      PrintUsage(argv[0]);
      return 1;
    } else if (address.empty()) {
      address = arg;
    } else if (model == "a") {
      model = arg;
    } else if (power == ".*") {
      power = arg;
    } else {
      PrintUsage(argv[0]);
      return 1;
    }
  }

  if (address.empty()) {
    PrintUsage(argv[0]);
    return 1;
  }

  model = ToLower(model);

  if (model == "a") {
    return RunRobotCollisions<y1_model::A>(address, power);
  }
  if (model == "m") {
    return RunRobotCollisions<y1_model::M>(address, power);
  }
  if (model == "ub") {
    return RunRobotCollisions<y1_model::UB>(address, power);
  }

  std::cerr << "Unknown model: " << model << std::endl;
  PrintUsage(argv[0]);
  return 1;
}
